drivers/rescontainervbox/main.go

```diff
@@ -0,0 +1,752 @@
+	}
+	doc, err := xmlquery.Parse(f)
+	if err != nil {
+		return false, err
+	}
+	machineEntries, err := xmlquery.QueryAll(doc, "//MachineEntry")
+	if err != nil {
+		return false, err
+	}
+	for _, entry := range machineEntries {
+		src := entry.SelectAttr("src")
+		if src == filepath.Join("/root/VirtualBox VMs", t.Name, t.Name+".vbox") {
+			return true, nil
+		}
+	}
+	return false, nil
+}
+
+func (t T) addVmToVboxCf() error {
+	cmd := command.New(
+		command.WithName("VBoxManage"),
+		command.WithVarArgs("registervm", filepath.Join("/root/VirtualBox VMs", t.Name, t.Name+".vbox")),
+		command.WithLogger(t.Log()),
+		command.WithCommandLogLevel(zerolog.InfoLevel),
+		command.WithStdoutLogLevel(zerolog.InfoLevel),
+		command.WithStderrLogLevel(zerolog.ErrorLevel),
+	)
+
+	return cmd.Run()
+}
+
+func (t *T) Start(ctx context.Context) error {
+	if v, err := t.isVmInVboxCf(); err != nil {
+		return err
+	} else if !v {
+		if err := t.addVmToVboxCf(); err != nil {
+			return err
+		}
+	}
+	if err := t.ApplyPGChain(ctx); err != nil {
+		return err
+	}
+	if v, err := t.isUp(); err != nil {
+		return err
+	} else if v {
+		t.Log().Info().Msgf("container %s is already up", t.Name)
+		return nil
+	}
+	if err := t.containerStart(ctx); err != nil {
+		return err
+	}
+	actionrollback.Register(ctx, func() error {
+		return t.Stop(ctx)
+	})
+	if err := t.waitForUp(); err != nil {
+		return err
+	}
+
+	if _, err := net.LookupIP(t.hostname()); err == nil {
+		if err := t.waitForPing(); err != nil {
+			return err
+		}
+		if err := t.waitForOperational(); err != nil {
+			return err
+		}
+	} else {
+		t.Log().Debug().Msgf("can not do dns resolution for : %s", t.Name)
+	}
+	return nil
+}
+
+func (t T) Stop(ctx context.Context) error {
+	if v, err := t.isDown(); err != nil {
+		return err
+	} else if v {
+		t.Log().Info().Msgf("container %s is already down", t.Name)
+		return nil
+	}
+	if err := t.containerStop(ctx); err != nil {
+		return err
+	}
+	return nil
+}
+
+func (t T) waitForDown() error {
+	t.Log().Info().Dur("timeout", *t.StopTimeout).Msgf("wait for %s shutdown, max duration %s", t.Name, *t.StopTimeout)
+	return WaitFor(func() bool {
+		v, err := t.isDown()
+		if err != nil {
+			return true
+		}
+		return v
+	}, time.Second*2, *t.StopTimeout)
+}
+
+func (t T) waitForUp() error {
+	t.Log().Info().Dur("timeout", *t.StartTimeout).Msgf("wait for %s up", t.Name)
+	return WaitFor(func() bool {
+		v, err := t.isUp()
+		if err != nil {
+			t.Log().Error().Err(err).Msgf("abort waiting for %s up", t.Name)
+			return true
+		}
+		return v
+	}, time.Second*2, *t.StopTimeout)
+}
+
+func (t T) waitForPing() error {
+	t.Log().Info().Dur("timeout", *t.StartTimeout).Msgf("wait for %s ping", t.Name)
+	return WaitFor(func() bool {
+		v, err := t.isPinging()
+		if err != nil {
+			t.Log().Error().Err(err).Msgf("abort waiting for %s ping", t.Name)
+			return true
+		}
+		return v
+	}, time.Second*2, *t.StopTimeout)
+}
+
+func (t T) waitForOperational() error {
```

Comment:
dup code, refactor waitForUp, Down, Ping, Operational to
func (t *T) waitForOperational(name, function() error) error {